text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,278 @@
+brings the user to a package detail page that represents the following
+information:
+
+- Package name, tags, and description
+- Overview of publishing information (author, version, license, etc...)
+  - Where possible, links to the source repository, issue tracker, etc...
+- Install instructions
+- Available languages
+- Package's documentation
+  - Rendered content of the package's `README.md`
+  - API reference documentation, in all supported languages
+
+The list of available languages is also a selector: upon clicking a language,
+the installation instructions and documentation automatically switch to the
+relevant form for the currently selected language.
+
+### Support
+
+The CDK Construct Hub offers ways for customers to engage maintainers about
+various problems, including the following:
+
+- A malicious package was identified
+- Incorrectly rendered documentation
+- Package un-listing request
+
+When a contact request originates from a package detail page, the customer is
+guided to the package's issue tracker or public repository if their issue is
+with the package itself (and not about it's listing in CDK Construct Hub).
+
+## Technical Design
+
+### What is the high level implementation plan?
+
+#### Overview
+
+The CDK Construct Hub is a relatively simple static web application. While the
+dataset it exposes (all CDK Construct Libraries) may be relatively fast moving,
+there is no necessity for newly published packages (or versions thereof) to be
+indexed and presented on the application particularly quickly. A consistency
+window between a couple of hours and a day is probably acceptable.
+
+A simple data pipeline will ingest package publication events created by a
+source-specific adapter (e.g: by polling on the CouchDB instance replica for the
+[npmjs.com registry](https://skimdb.npmjs.com/registry)), and produce static
+artifacts the website is based off of. This pipeline is in particular
+responsible for extracting the `.jsii` assembly file from the npm packages, and
+to prefetch all information necessary to back the webiste's search and package
+detail pages.
+
+#### Back-End
+
+1. The back-end for CDK Construct Hub starts with a custom-built event source,
+   which implements the logic necessary to detect new packages that are relevant
+   to the CDK Construct Hub, and sends messages to an SQS queue for further
+   processing. Notifications have the following attributes:
+
+   Name        | Description
+   ------------|----------------------------------------------------------------
+   `assembly`  | The `.jsii` assembly included in the package
+   `time`      | The timestamp at which the version was created
+   `integrity` | An integirity check for the complete record
+
+1. A Lambda function then picks up messages from the SQS queue and prepares the
+   artifacts consumed by the front-end application, stored in a dedicated S3
+   bucket using the following key format:
```

Comment:
Is this bucket also fronted by CloudFront? I believe the prototype does that.

Reply:
Yes - the app would be 100% served by CloudFront.